### PR TITLE
V0.4 implement keyword storage in logic history

### PR DESCRIPTION
**Pull request by kevinLamKB**
Added new component in CommandHistory that stores prev keyword search, complete with getters and setters.

- update find,list, and failed edit/delete by find commands to update prevKeyword and prevState in history
- update edit, delete to refresh by resetting filtered lists by previous keywords instead of setting them to showAll

@ChuaPingChan, @akd13 so for complete and restore commands, take note of the places to update. for both variants by index and by search, call the 3 updatelist methods that i marked with comments instead of previously where they called updateFilteredListShowAll, and for byFind 2 places to insert the statement to save the search terms and state. Lastly, override setData method in the Commands itself so that the command has a history to work with instead of null pointers.

### CODE REVIEW

**Comment by akd13** [comment]
Noted, thanks a lot!

File: src/main/java/seedu/multitasky/logic/CommandHistory.java [hunk 1]
```diff
@@ -30,4 +39,18 @@ public void add(String userInput) {
...
+    }
+
+    public State getPrevState() {
+        return previousState;
+    }
+
+    public void setNextSearch(Set<String> nextSearch, State nextState) {
+        previousSearch = nextSearch;
+        previousState = nextState;
+    }
```
Comment (ChuaPingChan) [inline src/main/java/seedu/multitasky/logic/CommandHistory.java]: `setNextSearch` seems a bit misleading to me, maybe `setPrevSearch` is better. What do you think?

File: src/main/java/seedu/multitasky/logic/CommandHistory.java [hunk 2]
```diff
@@ -3,16 +3,25 @@
...
+import seedu.multitasky.model.entry.Entry.State;
 
 /**
- * Stores the history of commands executed.
+ * Stores the history of commands executed, and also the keywords used in the previous filters of inner lists.
  */
 public class CommandHistory {
     private ArrayList<String> userInputHistory;
+    private Set<String> previousSearch;
+    private State previousState;
```
Comment (ChuaPingChan) [inline src/main/java/seedu/multitasky/logic/CommandHistory.java]: Suggestion: `previousKeywordsSearch` would be more descriptive than `previousSearch`.

**Review by ChuaPingChan — APPROVED** [review]
@kevinLamKB 
Noted with thanks. I added some suggestions.

File: src/main/java/seedu/multitasky/logic/CommandHistory.java (hunk 1 above)
Comment (kevinLamKB) [reply src/main/java/seedu/multitasky/logic/CommandHistory.java]: haha , thats actually what i named it as initially. i forgot why i changed it to next search

**Review by mattheuslee — APPROVED** [review]

**Comment by kevinLamKB** [comment]
Updated to setPrevSearch, merging now so that u guys can work on V0.4 side

File: src/main/java/seedu/multitasky/logic/commands/DeleteByIndexCommand.java [hunk 3]
```diff
@@ -41,7 +41,11 @@ public CommandResult execute() throws CommandException, DuplicateEntryException
         } catch (EntryNotFoundException enfe) {
             assert false : "The target entry cannot be missing";
         }
-        model.updateAllFilteredListToShowAllActiveEntries();
+        // refresh list view after updating.
+        model.updateFilteredDeadlineList(history.getPrevSearch(), Entry.State.ACTIVE);
+        model.updateFilteredEventList(history.getPrevSearch(), Entry.State.ACTIVE);
+        model.updateFilteredFloatingTaskList(history.getPrevSearch(), Entry.State.ACTIVE);
+
```
Comment (ChuaPingChan) [inline src/main/java/seedu/multitasky/logic/commands/DeleteByIndexCommand.java:48]: @kevinLamKB 
Hmm, should the second argument here be `history.getPrevState()` as how it is in `DeleteByFindCommand.java`?

**Pull request merged**